pkg/controller/contrailcni/contrailcni_controller.go

```diff
@@ -0,0 +1,224 @@
+func NewReconciler(client client.Client, scheme *runtime.Scheme, clusterInfo v1alpha1.VrouterClusterInfo) reconcile.Reconciler {
+	return &ReconcileContrailCNI{Client: client, Scheme: scheme, ClusterInfo: clusterInfo}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("contrailcni-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource ContrailCNI
+	err = c.Watch(&source.Kind{Type: &contrailv1alpha1.ContrailCNI{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to PODs.
+	serviceMap := map[string]string{"contrail_manager": "vrouter"}
+	srcPod := &source.Kind{Type: &corev1.Pod{}}
+	podHandler := resourceHandler(mgr.GetClient())
+	predInitStatus := utils.PodInitStatusChange(serviceMap)
+	predPodIPChange := utils.PodIPChange(serviceMap)
+	predInitRunning := utils.PodInitRunning(serviceMap)
+
+	if err = c.Watch(srcPod, podHandler, predPodIPChange); err != nil {
+		return err
+	}
+	if err = c.Watch(srcPod, podHandler, predInitStatus); err != nil {
+		return err
+	}
+	if err = c.Watch(srcPod, podHandler, predInitRunning); err != nil {
+		return err
+	}
+
+	srcDS := &source.Kind{Type: &appsv1.DaemonSet{}}
+	dsHandler := &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &v1alpha1.ContrailCNI{},
+	}
+	dsPred := utils.DSStatusChange(utils.ContrailCNIGroupKind())
+	if err = c.Watch(srcDS, dsHandler, dsPred); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileContrailCNI implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileContrailCNI{}
+
+// ReconcileContrailCNI reconciles a ContrailCNI object
+type ReconcileContrailCNI struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	Client      client.Client
+	Scheme      *runtime.Scheme
+	ClusterInfo v1alpha1.VrouterClusterInfo
+}
+
+// Reconcile reads that state of the cluster for a ContrailCNI object and makes changes based on the state read
+// and what is in the ContrailCNI.Spec
+func (r *ReconcileContrailCNI) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling ContrailCNI")
+	instanceType := "contrailcni"
+	instance := &v1alpha1.ContrailCNI{}
+	ctx := context.TODO()
+
+	if err := r.Client.Get(ctx, request.NamespacedName, instance); err != nil && errors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	}
+
+	if !instance.GetDeletionTimestamp().IsZero() {
+		return reconcile.Result{}, nil
+	}
+
+	_, err := instance.CreateConfigMap(request.Name+"-"+instanceType+"-configuration", r.Client, r.Scheme, request)
```

Comment:
Please move creation of config maps to this controller like in openstack services or command

Reply:
Done